File: week09/20/pr1_find_if_path_exists_in_graph.cpp
using namespace std;
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        unordered_map<int,vector<int>> graph;
        for(auto vec : edges){
            graph[vec[0]].push_back(vec[1]);
            graph[vec[1]].push_back(vec[0]);
        }
        
        queue<int> q;
        unordered_set<int> visited;

        q.push(source);
        visited.insert(source);

        while(!q.empty()){
            int node = q.front();
            q.pop();
            if(node == destination) return true;

            for(int x : graph[node]){
                if(visited.find(x) == visited.end()){
                    visited.insert(x);
                    q.push(x);}
            }
        }
        return false;
    }
};

File: week09/20/pr1_find_if_path_exists_in_graph.py
class Solution:
    def validPath(self, n: int, edges: list[list[int]], source: int, destination: int) -> bool:
        # make adjacent list
        g = {i:[] for i in range(n)} #dictionary
        for u,v in edges:
            g[u] += [v]
            g[v] += [u]
        
        visited = set()

        # Recursive
        def helper(node):
            if node == destination: return True # 아..함수 안에 사용하면 바깥 함수 varialbe 사용가능하네
            visited.add(node)
            print(visited)
            for neighbor in g[node]:
                if neighbor not in visited:
                    if helper(neighbor):
                        return True
            return False
        return helper(source)

s = Solution()
n = 6
edges = [[0,1],[0,2],[3,5],[5,4],[4,3]]
print(s.validPath(n,edges,0,5))